src/compiler/checker.ts

```diff
@@ -30362,8 +30362,8 @@ namespace ts {
             else if (type && !hasExplicitReturn) {
                 // minimal check: function has syntactic return type annotation and no explicit return statements in the body
                 // this function does not conform to the specification.
-                // NOTE: having returnType !== undefined is a precondition for entering this branch so func.type will always be present
-                error(getEffectiveReturnTypeNode(func), Diagnostics.A_function_whose_declared_type_is_neither_void_nor_any_must_return_a_value);
+                // NOTE: returnType may not exist when func.type is denoted by jsdoc
+                error(getEffectiveReturnTypeNode(func) || func, Diagnostics.A_function_whose_declared_type_is_neither_void_nor_any_must_return_a_value);

```

Comment:
all 4 four branches of this if at the end of the function now have `getEffectiveReturnTypeNode(func) || func`, so I'd lift it out before the first if as `const errorNode = getEffectiveReturnTypeNode(func) || func`

If you can, you should find or write a new test for the `type.flags & TypeFlags.Never` case above as well.

Reply:
👌 👌 Got it.